### PR TITLE
ET0002: Add CRUD operations for Taxes  

### PR DESCRIPTION
**Pull request by EvelinaTemnolonskyte**

### CODE REVIEW

File: src/main/java/com/example/demo/taxComponent/api/dtos/GetTaxesDTO.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    private int totalItems;
+    private int totalPages;
+    private int currentPage;
+    private List<TaxDTO> items;
+
+    public GetTaxesDTO(List<TaxDTO> taxes, int totalItems) {
+        this.items = taxes;
+        this.totalItems = totalItems;
+    }
+}
```
Comment (d-pilipavicius) [inline src/main/java/com/example/demo/taxComponent/api/dtos/GetTaxesDTO.java]: Add a @Builder annotation, so you wouldn't need a constructor, make all fields final, remove @Setter, @NoArgsConstructor

File: src/main/java/com/example/demo/taxComponent/api/dtos/PostTaxDTO.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+public class PostTaxDTO {
+    private String title;
+    private BigDecimal ratePercentage;
+    private UUID businessId;
+}
```
Comment (d-pilipavicius) [inline src/main/java/com/example/demo/taxComponent/api/dtos/PostTaxDTO.java]: Remove @Setter, all fields final, remove @NoArgsConstructror

File: src/main/java/com/example/demo/taxComponent/api/dtos/PutTaxDTO.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import java.util.Optional;
+
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+public class PutTaxDTO {
+    private Optional<String> title;
+    private Optional<BigDecimal> ratePercentage;
+}
```
Comment (d-pilipavicius) [inline src/main/java/com/example/demo/taxComponent/api/dtos/PutTaxDTO.java]: Remove @Setter, all fields final, remove @NoArgsConstructror

File: src/main/java/com/example/demo/taxComponent/api/dtos/ResponseTaxDTO.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+/**
+ * A response tax DTO for Patch and Post requests
+ */
+public class ResponseTaxDTO {
+    private TaxDTO tax;
```
Comment (d-pilipavicius) [inline src/main/java/com/example/demo/taxComponent/api/dtos/ResponseTaxDTO.java]: Remove this object, rewrite code to use TaxDTO

File: src/main/java/com/example/demo/taxComponent/api/dtos/TaxHelperDTOs/TaxDTO.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+public class TaxDTO {
+    private UUID id;
+    private String title;
+    private BigDecimal ratePercentage;
+    private UUID businessId;
+}
```
Comment (d-pilipavicius) [inline src/main/java/com/example/demo/taxComponent/api/dtos/TaxHelperDTOs/TaxDTO.java]: Remove @Setter, all fields final, remove @NoArgsConstructror

File: src/main/java/com/example/demo/taxComponent/api/endpoints/TaxesController.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+@RestController
+@RequestMapping("/v1/taxes")
+public class TaxesController {
+
+    private TaxApplicationService taxApplicationService;
+
+    @Autowired
+    public TaxesController(TaxApplicationService taxApplicationService){
+        this.taxApplicationService = taxApplicationService;
+    }
```
Comment (d-pilipavicius) [inline src/main/java/com/example/demo/taxComponent/api/endpoints/TaxesController.java]: Use @AllArgsConstructor instead

File: src/main/java/com/example/demo/taxComponent/api/endpoints/TaxesController.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    public ResponseEntity<Object> deleteTax(@PathVariable UUID taxId){
+        taxApplicationService.deleteTax(taxId);
+        return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
+    }
+
+    @GetMapping
+    public ResponseEntity<Object> getTaxes(@RequestParam int pageNumber, @RequestParam int pageSize){
+        GetTaxesDTO taxes = taxApplicationService.getAllTaxes(pageNumber, pageSize);
+        return ResponseEntity.status(HttpStatus.OK).body(taxes);
+    }
```
Comment (d-pilipavicius) [inline src/main/java/com/example/demo/taxComponent/api/endpoints/TaxesController.java]: Add @NotNull annotation on such objects as UUID, dto's

File: src/main/java/com/example/demo/taxComponent/applicationServices/TaxApplicationService.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+@Service
+public class TaxApplicationService {
+
+    private static final Logger logger = LoggerFactory.getLogger(TaxApplicationService.class);
+
+    private final TaxService taxService;
+
+    public TaxApplicationService(TaxService taxService) {
+        this.taxService = taxService;
+    }
```
Comment (d-pilipavicius) [inline src/main/java/com/example/demo/taxComponent/applicationServices/TaxApplicationService.java]: @AllArgsConstructor

File: src/main/java/com/example/demo/taxComponent/api/endpoints/TaxesController.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+import java.util.UUID;
+
+
+
+@RestController
+@RequestMapping("/v1/taxes")
+public class TaxesController {
+
+    private TaxApplicationService taxApplicationService;
```
Comment (d-pilipavicius) [inline src/main/java/com/example/demo/taxComponent/api/endpoints/TaxesController.java]: Field should be final

File: src/main/java/com/example/demo/taxComponent/applicationServices/TaxApplicationService.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+    @Transactional
+    public TaxDTO createTax(PostTaxDTO postTaxDTO) {
+        return taxService.createTax(postTaxDTO);
+    }
+
+    public GetTaxesDTO getAllTaxes() {
+        return taxService.getAllTaxes();
+    }
```
Comment (d-pilipavicius) [inline src/main/java/com/example/demo/taxComponent/applicationServices/TaxApplicationService.java]: This method will probably never get used, maybe remove it?

File: src/main/java/com/example/demo/taxComponent/domain/entities/Tax.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+import lombok.NoArgsConstructor;
+
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.UUID;
+
+@Entity
+@AllArgsConstructor
+@NoArgsConstructor
+@Data
```
Comment (d-pilipavicius) [inline src/main/java/com/example/demo/taxComponent/domain/entities/Tax.java]: Change data to @Getter, @Setter

File: src/main/java/com/example/demo/taxComponent/domain/entities/Tax.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    @Column(name = "business_id", nullable = false)
+    private UUID businessId;
+
+   /* @ManyToMany(mappedBy = "taxes")
+    private List<Product> products; */
+
+    @Column(nullable = false)
+    private String title;
+
+    @Column(nullable = false, precision = 10, scale = 2)
```
Comment (d-pilipavicius) [inline src/main/java/com/example/demo/taxComponent/domain/entities/Tax.java]: Give name in @Column that is snake case version (add name = "rate_percentage")

File: src/main/java/com/example/demo/taxComponent/domain/services/TaxService.java [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+@Service
+public class TaxService {
+
+    private static final Logger logger = LoggerFactory.getLogger(TaxService.class);
+
+    private final TaxRepository taxRepository;
+
+    public TaxService(TaxRepository taxRepository) {
+        this.taxRepository = taxRepository;
+    }
```
Comment (d-pilipavicius) [inline src/main/java/com/example/demo/taxComponent/domain/services/TaxService.java]: @AllArgsConstructor

File: src/main/java/com/example/demo/taxComponent/domain/services/TaxService.java [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+        );
+    }
+
+    public GetTaxesDTO getAllTaxes(){
+        List<TaxDTO> taxes = taxRepository.findAll().stream()
+                .map(TaxMapper.TO_DTO::map)
+                .collect(Collectors.toList());
+
+        return new GetTaxesDTO(taxes, taxes.size());
+    }
```
Comment (d-pilipavicius) [inline src/main/java/com/example/demo/taxComponent/domain/services/TaxService.java]: Method probably will not be used

File: src/main/java/com/example/demo/taxComponent/helper/mapper/TaxMapper.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        return Mapper.mapToDTOList(entities, TO_DTO);
+    }
+
+    public static Optional<Tax> mapToModelOptional(PostTaxDTO dto) {
+        return Mapper.mapToModelOptional(dto, TO_MODEL);
+    }
+
+    public static Optional<TaxDTO> mapToDTOOptional(Tax entity) {
+        return Mapper.mapToDTOOptional(entity, TO_DTO);
+    }
```
Comment (d-pilipavicius) [inline src/main/java/com/example/demo/taxComponent/helper/mapper/TaxMapper.java]: If you are using this mapper, adhere to the variable system, not method.

**Review by d-pilipavicius — DISMISSED** [review]
LGTM (Remove all @Transactions annotations if you do not use them next time)

**Comment by d-pilipavicius** [comment]
LGTM

**Review by d-pilipavicius — DISMISSED** [review]
LGMT

**Comment by d-pilipavicius** [comment]
LGTM

**Review by migle-kirilovaite — APPROVED** [review]

**Review by VarArnas — APPROVED** [review]

**Pull request merged**